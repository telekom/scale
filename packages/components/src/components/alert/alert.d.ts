export declare class Alert {
    /** (required) Alert class */
    customClass?: string;
    /** (optional) Alert size */
    size?: string;
    /** (optional) Alert theme */
    theme?: string;
    /** (optional) Alert variant */
    variant?: string;
    /** (optional) Alert title */
    headline: string;
    /** (required) Alert opened */
    opened: boolean;
    /** (optional) Alert timeout */
    timeout?: boolean | number;
    /** (optional) Alert icon */
    icon?: string;
    /** (required) Alert close */
    close?: string;
    private getCssClassMap;
    private defaultTimeout;
    onCloseAlert: () => void;
    open(): Promise<void>;
    onCloseAlertWithTimeout: () => any;
    render(): any;
}
