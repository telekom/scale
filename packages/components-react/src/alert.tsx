import * as React from 'react'
require('@telements/styles/dist/alert.css');

interface Props {
	children?: any;
	context?: string;
}

export const Alert: React.SFC<Props> = ({ children, context }) => (
	<div className={`alert ${context !== '' ? `alert--${context}` : ''}`}>
		{children}
	</div>
)
