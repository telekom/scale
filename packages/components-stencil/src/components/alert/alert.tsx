import { Component, Prop, h } from '@stencil/core';
import { CssClassMap } from '../../utils/utils';
import classNames from 'classnames';

@Component({
  tag: 't-alert',
  styleUrls: [
    './../../../../../node_modules/@telements/styles/dist/alert.css',
    'alert.css'
  ],
  shadow: true
})
export class alert {
  @Prop() customClass?: string = '';
  @Prop() size?: string = '';
  @Prop() theme?: string = '';
  @Prop() variant?: string = '';

  private getCssClassMap(): CssClassMap {
    return classNames(
      'alert',
      this.customClass && this.customClass,
      this.size && `alert--size-${this.size}`,
      this.theme && `alert--theme-${this.theme}`,
      this.variant && `alert--variant-${this.variant}`,
    );
  }

  render() {
    return (
      <div class={this.getCssClassMap()}>
        <slot/>
      </div>
    );
  }
}
