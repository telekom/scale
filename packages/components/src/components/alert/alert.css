/**
 * @prop --alert-background: Background color of the alert
 * @prop --alert-color: Color of the text of the alert
 * @prop --alert-padding: Padding of the alert
 * @prop --alert-headline-font-size: Font size of the headline of the alert
 * @prop --alert-headline-color: Color of the headline of the alert
 * @prop --alert-headline-margin: Margin of the headline of the alert
 * @prop --alert-icon-background: Background color of the icon of the alert
 * @prop --alert-icon-height: Height of the icon of the alert
 * @prop --alert-icon-width: Width of the icon of the alert
 * @prop --alert-icon-margin: Margin of the icon of the alert
 * @prop --alert-icon-border-radius: Border radius of the icon of the alert
 * @prop --alert-close-width: Width of the close icon of the alert
 * @prop --alert-close-height: Height of the close icon of the alert
 * @prop --alert-close-opacity: Opacity of the close icon of the alert
 * @prop --alert-close-hover-opacity: Hover opacity of the close icon of the alert
 * @prop --alert-variant-primary-background: Background color of the variant primary alert
 * @prop --alert-variant-primary-color: Color of the text the variant primary alert
 * @prop --alert-variant-secondary-background: Background color of the variant secondary alert
 * @prop --alert-variant-secondary-color: Color of the text of the variant secondary alert
 * @prop --alert-variant-warning-background: Background color of the variant warning alert
 * @prop --alert-variant-warning-color: Color of the text of the variant warning alert
 * @prop --alert-variant-danger-background: Background color of the variant danger alert
 * @prop --alert-variant-danger-color: Color of the text of the variant danger alert
 * @prop --alert-variant-success-background: Background color of the variant success alert
 * @prop --alert-variant-success-color: Color of the text of the variant success alert
 * @prop --alert-variant-info-background: Background color of the variant info alert
 * @prop --alert-variant-info-color: Color of the text of the variant info alert
 */

.alert {
  box-sizing: border-box;
  background: var(--alert-background, #eee);
  color: var(--alert-color, #333);
  width: 100%;
  padding: var(--alert-padding, 1rem);
  text-align: left;
  position: relative;
  display: flex;
  justify-content: space-between;
}

.alert__body {
  display: flex;
}

.alert__headline {
  font-size: var(--alert-headline-font-size, 1rem);
  color: var(--alert-headline-color, white);
  margin: var(--alert-headline-margin, 0);
}

.alert__icon {
  display: flex;
  align-items: center;
  justify-content: center;
  background: var(--alert-icon-background);
  height: var(--alert-icon-height, 24px);
  width: var(--alert-icon-width, 24px);
  margin: var(--alert-icon-margin, 0 0.5rem 0 0);
  border-radius: var(--alert-icon-border-radius);
}

.alert__close {
  width: var(--alert-close-width, 16px);
  height: var(--alert-close-height, 16px);
  opacity: var(--alert-close-opacity, 0.5);
  cursor: pointer;
}

.alert__close:hover {
  opacity: var(--alert-close-hover-opacity, 1);
}

.alert--variant-primary {
  background: var(--alert-variant-primary-background, blue);
  color: var(--alert-variant-primary-color, #fff);
}

.alert--variant-secondary {
  background: var(--alert-variant-secondary-background, #eee);
  color: var(--alert-variant-secondary-color, #333);
}

.alert--variant-variant-warning {
  background: var(--alert-variant-warning-background, orange);
  color: var(--alert-variant-warning-color, #fff);
}

.alert--variant-danger {
  background: var(--alert-variant-danger-background, red);
  color: var(--alert-variant-danger-color, #fff);
}

.alert--variant-success {
  background: var(--alert-variant-success-background, green);
  color: var(--alert-variant-success-color, #fff);
}

.alert--variant-info {
  background: var(--alert-variant-info-background, lightblue);
  color: var(--alert-variant-info-color, #fff);
}
