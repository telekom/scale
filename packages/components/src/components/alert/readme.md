# t-alert



<!-- Auto Generated Below -->


## Properties

| Property      | Attribute      | Description              | Type                | Default     |
| ------------- | -------------- | ------------------------ | ------------------- | ----------- |
| `customClass` | `custom-class` | (optional) Alert class   | `string`            | `''`        |
| `headline`    | `headline`     | (optional) Alert title   | `string`            | `undefined` |
| `icon`        | `icon`         | (optional) Alert icon    | `string`            | `''`        |
| `opened`      | `opened`       | (optional) Alert opened  | `boolean`           | `undefined` |
| `size`        | `size`         | (optional) Alert size    | `string`            | `''`        |
| `theme`       | `theme`        | (optional) Alert theme   | `string`            | `''`        |
| `timeout`     | `timeout`      | (optional) Alert timeout | `boolean \| number` | `false`     |
| `variant`     | `variant`      | (optional) Alert variant | `string`            | `''`        |


## Methods

### `open() => Promise<void>`

Alert method: open()

#### Returns

Type: `Promise<void>`




## CSS Custom Properties

| Name                                   | Description                                      |
| -------------------------------------- | ------------------------------------------------ |
| `--alert-background`                   | Background color of the alert                    |
| `--alert-close-height`                 | Height of the close icon of the alert            |
| `--alert-close-hover-opacity`          | Hover opacity of the close icon of the alert     |
| `--alert-close-opacity`                | Opacity of the close icon of the alert           |
| `--alert-close-width`                  | Width of the close icon of the alert             |
| `--alert-color`                        | Color of the text of the alert                   |
| `--alert-headline-color`               | Color of the headline of the alert               |
| `--alert-headline-font-size`           | Font size of the headline of the alert           |
| `--alert-headline-margin`              | Margin of the headline of the alert              |
| `--alert-icon-background`              | Background color of the icon of the alert        |
| `--alert-icon-border-radius`           | Border radius of the icon of the alert           |
| `--alert-icon-height`                  | Height of the icon of the alert                  |
| `--alert-icon-margin`                  | Margin of the icon of the alert                  |
| `--alert-icon-width`                   | Width of the icon of the alert                   |
| `--alert-padding`                      | Padding of the alert                             |
| `--alert-variant-danger-background`    | Background color of the variant danger alert     |
| `--alert-variant-danger-color`         | Color of the text of the variant danger alert    |
| `--alert-variant-info-background`      | Background color of the variant info alert       |
| `--alert-variant-info-color`           | Color of the text of the variant info alert      |
| `--alert-variant-primary-background`   | Background color of the variant primary alert    |
| `--alert-variant-primary-color`        | Color of the text the variant primary alert      |
| `--alert-variant-secondary-background` | Background color of the variant secondary alert  |
| `--alert-variant-secondary-color`      | Color of the text of the variant secondary alert |
| `--alert-variant-success-background`   | Background color of the variant success alert    |
| `--alert-variant-success-color`        | Color of the text of the variant success alert   |
| `--alert-variant-warning-background`   | Background color of the variant warning alert    |
| `--alert-variant-warning-color`        | Color of the text of the variant warning alert   |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
