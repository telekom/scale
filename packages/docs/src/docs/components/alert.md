---
title: Alert
description: Alert element
url: /docs/alert
contributors:
  - eeegor
---

# Alert

Alert component

selector: `t-alert`

## Default

```javascript
 <t-button id="show-alert-btn"
      >show alert
    </t-button>
    <t-alert id="alert" variant="danger" title="Help!" icon="!">
      This is only a warning - for your information!
      <div slot="close">Close</div>
    </t-alert>
    <script>
      const showAlert = document.querySelector(
        '#show-alert-btn'
      );
      const alert = document.getElementById('alert');
      showAlert.addEventListener(
        'click',
        () => !alert.opened && alert.open()
      );
    </script>
```

<div class="demo-container">
 <t-button id="show-alert-btn-without-timeout"
      >show alert
    </t-button>
    <t-alert id="alert-without-timeout" variant="danger" title="Help!" icon="!">
      This is only a warning - for your information!
      <div slot="close">Close</div>
    </t-alert>
    <script>
      const showAlertWithout = document.querySelector(
        '#show-alert-btn-without-timeout'
      );
      const alertWithout = document.getElementById('alert-without-timeout');
      showAlertWithout.addEventListener(
        'click',
        () => !alertWithout.opened && alertWithout.open()
      );
    </script>
</div>

## Opened

```javascript
<t-alert opened="true">This is only a warning - for your information!</t-alert>
```

<div class="demo-container">
  <t-alert opened="true">
    This is only a warning - for your information!
</t-alert>
</div>

## Variants

Available `variants`:

- danger
- success
- warning
- info
- primary
- secondary

```javascript
<t-alert opened="true" variant="danger">
  This is only a warning - for your information!
</t-alert>
```

<div class="demo-container">
  <t-alert opened="true" variant="danger">
    This is only a warning - for your information!
  </t-alert>
</div>
