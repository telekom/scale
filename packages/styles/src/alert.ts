import { css } from "./util";

export const alert = css`
	.alert {
		box-sizing: border-box;
		background: #eee;
		color: #333;
		width: 100%;
		padding: 1rem;
		text-align: left;
	}

	.alert--primary {
		background: blue;
		color: #fff;
	}

	.alert--secondary {
		background: #eee;
		color: #333;
	}

	.alert--warning {
		background: orange;
		color: #fff;
	}

	.alert--danger {
		background: red;
		color: #fff;
	}

	.alert--success {
		background: green;
		color: #fff;
	}

	.alert--info {
		background: lightblue;
		color: #fff;
	}
`
