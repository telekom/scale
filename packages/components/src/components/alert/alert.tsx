import { Component, Prop, h, Method, Element, Host } from '@stencil/core';
import { HTMLStencilElement } from '@stencil/core/internal';
import { CssClassMap } from '../../utils/utils';
import classNames from 'classnames';

import { styles } from './alert.styles';
import { CssInJs } from '../../utils/css-in-js';
import { StyleSheet } from 'jss';
import Base from '../../utils/base-interface';

@Component({
  tag: 'scale-alert',
  shadow: true,
})
export class Alert implements Base {
  @Element() hostElement: HTMLStencilElement;
  /** (optional) Alert class */
  @Prop() customClass?: string = '';
  /** (optional) Alert size */
  @Prop() size?: string = '';
  /** (optional) Alert variant */
  @Prop() variant?: string = '';
  /** (optional) Alert title */
  @Prop({ reflectToAttr: true }) headline: string;
  /** (optional) Alert opened */
  @Prop({ reflectToAttr: true }) opened: boolean;
  /** (optional) Alert timeout */
  @Prop() timeout?: boolean | number = false;
  /** (optional) Alert icon */
  @Prop() icon?: string = '';

  /** (optional) Injected jss styles */
  @Prop() styles?: StyleSheet;
  /** decorator Jss stylesheet */
  @CssInJs('Alert', styles) stylesheet: StyleSheet;

  defaultTimeout = 2000;

  hasSlotClose: boolean;

  componentWillLoad() {
    this.hasSlotClose = !!this.hostElement.querySelector('[slot="close"]');
  }
  componentWillUpdate() {}
  componentDidUnload() {}

  close = () => {
    this.opened = false;
  };

  /** Alert method: open() */
  @Method()
  async open() {
    this.opened = true;
  }

  onCloseAlertWithTimeout = () => {
    if (this.timeout !== false) {
      if (typeof this.timeout === 'number') {
        setTimeout(this.close, this.timeout);
      } else {
        setTimeout(this.close, this.defaultTimeout);
      }
    } else {
      return null;
    }
  };

  render() {
    const { classes } = this.stylesheet;

    this.onCloseAlertWithTimeout();

    if (!this.opened) {
      return null;
    }

    return (
      <Host>
        <style>{this.stylesheet.toString()}</style>
        <div class={this.getCssClassMap()}>
          <div class={classes.alert__body}>
            <div class={classes.alert__icon}>{this.icon}</div>
            <div class={classes.alert__content}>
              <div class={classes.alert__headline}>{this.headline}</div>
              <slot />
            </div>
          </div>

          <a class={classes.alert__close} onClick={this.close}>
            {this.hasSlotClose ? (
              <div class={classes['alert__close-icon']}>
                <slot name="close" />
              </div>
            ) : (
              'x'
            )}
          </a>
        </div>
      </Host>
    );
  }

  getCssClassMap(): CssClassMap {
    const { classes } = this.stylesheet;
    return classNames(
      classes.alert,
      this.customClass && this.customClass,
      this.size && classes[`alert--size-${this.size}`],
      this.variant && classes[`alert--variant-${this.variant}`]
    );
  }
}
