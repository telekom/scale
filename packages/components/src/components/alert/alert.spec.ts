import { newSpecPage } from '@stencil/core/testing';
import { Alert } from './alert';

describe('Alert', () => {
  let element;
  beforeEach(async () => {
    element = new Alert();
    jest.useFakeTimers();
  });

  const components = [Alert];

  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components,
      html: `<t-alert>Notifications</t-alert>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should match snapshot when opened', async () => {
    const page = await newSpecPage({
      components,
      html: `<t-alert opened=true >Notifications</t-alert>`,
    });
    expect(page.root.shadowRoot).toBeTruthy();
    expect(page.root).toMatchSnapshot();
  });

  it('should close the alert', () => {
    expect(element.opened).toBe(undefined);
    element.close();
    expect(element.opened).toBe(false);
  });

  it('should open the alert', () => {
    expect(element.opened).toBe(undefined);
    element.open();
    expect(element.opened).toBe(true);
  });

  it('should not open the alert/ should not render, if the alert is already opened', () => {
    element.opened = true;
    expect(element.root).toBeFalsy();
  });

  it('should closed the alert after default timeout', () => {
    element.timeout = true;
    element.onCloseAlertWithTimeout();

    expect(setTimeout).toHaveBeenCalledTimes(1);
    expect(setTimeout).toHaveBeenLastCalledWith(
      expect.any(Function),
      element.defaultTimeout
    );
  });

  it('should closed the alert after set timeout', () => {
    element.timeout = 500;
    element.onCloseAlertWithTimeout();

    expect(setTimeout).toHaveBeenCalledTimes(1);
    expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 500);
  });

  it('should have a default css class', () => {
    expect(element.getCssClassMap()).toBe('alert');
  });

  it('should handle custom css class', () => {
    element.customClass = 'custom-class';
    expect(element.getCssClassMap()).toContain('custom-class');
  });

  it('should handle size css class', () => {
    element.size = 'small';
    expect(element.getCssClassMap()).toContain('alert--size-small');
  });

  it('should handle theme css class', () => {
    element.theme = 'default';
    expect(element.getCssClassMap()).toContain('alert--theme-default');
  });

  it('should handle variant css class', () => {
    element.variant = 'primary';
    expect(element.getCssClassMap()).toContain('alert--variant-primary');
  });
});
