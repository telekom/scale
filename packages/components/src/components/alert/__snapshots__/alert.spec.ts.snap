// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert should match snapshot 1`] = `
<t-alert>
  <mock:shadow-root></mock:shadow-root>
  Notifications
</t-alert>
`;

exports[`Alert should match snapshot when opened 1`] = `
<t-alert opened="">
  <mock:shadow-root>
    <div class="alert">
      <div class="alert__body">
        <div class="alert__icon"></div>
        <div class="alert__content">
          <div class="alert__headline"></div>
          <slot></slot>
        </div>
      </div>
      <a class="alert__close">
        x
      </a>
    </div>
  </mock:shadow-root>
  Notifications
</t-alert>
`;
